drivers/adc/ad7616/iio_ad7616.c

```diff
@@ -0,0 +1,327 @@
+	case AD7616_2V5:
+		return sprintf(buf, "2.5");
+	default:
+		return -EINVAL;
+	}
+}
+
+static int set_scale(void *device,
+		     char *buf,
+		     uint32_t len,
+		     const struct iio_ch_info *channel,
+		     intptr_t id)
+{
+	int ret;
+	int val;
+	enum ad7616_range range;
+	enum ad7616_ch ch = (channel->ch_num < 8) ? channel->ch_num : AD7616_VB0 +
+			    channel->ch_num - 8;
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	ret = sscanf(buf, "%u", &val);
+	if (ret != 1)
+		return -EINVAL;
+
+	switch (val) {
+	case 10:
+		range = AD7616_10V;
+		break;
+	case 5:
+		range = AD7616_5V;
+		break;
+	case 2:
+		range = AD7616_2V5;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return ad7616_set_range(dev, ch, range);
+}
+
+/**
+ * @brief	Read buffer data corresponding to AD7616 IIO device
+ * @param	iio_dev_data[in] - Pointer to IIO device data structure
+ * @return	0 in case of success, negative error code otherwise
+ */
+static int32_t iio_ad7616_submit_buffer(struct iio_device_data *iio_dev_data)
+{
+	uint16_t ret;
+	struct ad7616_conversion_result results[8];
+	uint16_t data[2];
+	uint32_t i = 0;
+	struct ad7616_iio_dev *iio_dev = iio_dev_data->dev;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	for (; i < iio_dev_data->buffer->samples; i++) {
+		ret = ad7616_read_data_serial(dev, results, 1);
+		if (ret)
+			return ret;
+
+		data[0] = results[0].channel_a;
+		data[1] = results[0].channel_b;
+
+		ret = iio_buffer_push_scan(iio_dev_data->buffer, data);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+
+static struct scan_type ad7616_iio_scan_type = {
+	.sign = 's',
+	.realbits = CHN_STORAGE_BITS,
+	.storagebits = CHN_STORAGE_BITS,
+	.shift = 0,
+	.is_big_endian = false
+};
+
+static struct iio_attribute ad7616_iio_ch_attributes[] = {
+	{ .name = "raw", .show = get_raw },
+	{ .name = "scale", .show = get_scale, .store = set_scale },
+	END_ATTRIBUTES_ARRAY
+};
+
+static struct iio_attribute ad7616_iio_global_attributes[] = {
+	AD7616_CHN_ATTR("crc", CRC_ATTR_ID),
+	END_ATTRIBUTES_ARRAY
+};
+
+static struct iio_channel ad7616_channels[] = {
+	AD7616_CH(0),
+	AD7616_CH(1),
+	AD7616_CH(2),
+	AD7616_CH(3),
+	AD7616_CH(4),
+	AD7616_CH(5),
+	AD7616_CH(6),
+	AD7616_CH(7),
+	AD7616_CH(8),
+	AD7616_CH(9),
+	AD7616_CH(10),
+	AD7616_CH(11),
+	AD7616_CH(12),
+	AD7616_CH(13),
+	AD7616_CH(14),
+	AD7616_CH(15),
+};
+
+static struct iio_device ad7616_iio_device_template = {
+	.num_ch = NO_OS_ARRAY_SIZE(ad7616_channels),
+	.channels = ad7616_channels,
+	.attributes = ad7616_iio_global_attributes,
+	.debug_reg_read = iio_ad7616_debug_reg_read,
+	.debug_reg_write= iio_ad7616_debug_reg_write,
+	.submit = iio_ad7616_submit_buffer,
+};
+
+/**
+ * @brief	Init for reading/writing and parameterization of a
+ *			ad7616 IIO device
+ * @param	desc[in,out] - IIO device descriptor
+ * @return	0 in case of success, negative error code otherwise
+ */
+int32_t ad7616_iio_param_init(struct iio_device **desc)
```

Comment:
is this API needed? the iio_init is already assigning the iio_dev structure

Reply:
No, sorry